File: src/ru/gurps/generator/desktop/models/rules/Spell.java
package ru.gurps.generator.desktop.models.rules;

import javafx.scene.control.Button;
import javafx.scene.control.TableRow;
import ru.gurps.generator.desktop.Main;
import ru.gurps.generator.desktop.config.Model;

public class Spell extends Model {
    public Integer id;
    public Integer schoolId;
    public String name;
    public String nameEn;
    public String spellType;
    public String description;
    public Integer complexity;
    public String cost;
    public String needTime;
    public String duration;
    public String maintainingCost;
    public String thing;
    public String createCost;
    public String demands;
    public String resistance;
    public String modifiers;
    @Ignore public Integer level = 1;
    @Ignore public Integer finalCost;
    @Ignore public Boolean add = false;
    @Ignore public Button addButton;
    @Ignore public Button removeButton;
    @Ignore public TableRow row;

    public Spell() {
    }

    public String getName() {
        return name;
    }

    public String getNameEn() {
        return nameEn;
    }

    public Integer getFinalCost() {
        return finalCost;
    }

    public Integer getLevel() {
        return level;
    }

    public String getSchool() {
        return ((School) new School().find(schoolId)).name;
    }

    public String getSchoolSingle() {
        return Main.locale.getString("school") + ": " + getSchool();
    }

    public String getResistanceSingle() {
        String res = resistance == null ? Main.locale.getString("no") : resistance;
        return Main.locale.getString("resistance") + ": " + res;
    }

    public String getTypeSingle() {
        return Main.locale.getString("type") + ": " + getSpellType();
    }

    public String getComplexitySingle() {
        return Main.locale.getString("complexity") + ": " + getComplexity();
    }

    public String getDemandsSingle() {
        String dem = demands == null ? Main.locale.getString("no") : demands;
        return Main.locale.getString("demands")+": " + dem;
    }

    public String getModifiersSingle() {
        if(modifiers == null) return "";
        else return Main.locale.getString("modifiers")+": " + modifiers;
    }

    public String getNeedTimeSingle() {
        return Main.locale.getString("need_time")+": " + needTime;
    }

    public String getCostSingle() {
        return Main.locale.getString("cost")+": " + cost;
    }

    public String getMaintainingCostSingle() {
        return Main.locale.getString("maintaining_cost")+": " + maintainingCost;
    }

    public String getDurationSingle() {
        return Main.locale.getString("duration")+": " + duration;
    }

    public String getThingSingle() {
        String thingS = thing == null ? Main.locale.getString("no") : thing;
        return Main.locale.getString("duration")+": " + thingS;
    }

    public String getCreateCostSingle() {
        String create = createCost == null ? Main.locale.getString("no") : createCost;
        return Main.locale.getString("duration")+": " + create;
    }

    public String getSpellType() {
        String new_type = spellType;
        new_type = new_type.replace(",", "; ");
        new_type = new_type.replace(":", Main.locale.getString("or")+ " ");
        new_type = new_type.replace("0", Main.locale.getString("usual") + " ");
        new_type = new_type.replace("1", Main.locale.getString("area") + " ");
        new_type = new_type.replace("2", Main.locale.getString("contact") + " ");
        new_type = new_type.replace("3", Main.locale.getString("throw") + " ");
        new_type = new_type.replace("4", Main.locale.getString("block_spell") + " ");
        new_type = new_type.replace("5", Main.locale.getString("resistance") + " ");
        new_type = new_type.replace("6", Main.locale.getString("information") + " ");
        new_type = new_type.replace("7", Main.locale.getString("charm") + " ");
        new_type = new_type.replace("8", Main.locale.getString("special") + " ");
        return new_type;
    }

    public String getComplexity() {
        switch(complexity){
            case 0: return Main.locale.getString("easy");
            case 1: return Main.locale.getString("medium");
            case 2: return Main.locale.getString("hard");
            case 3: return Main.locale.getString("very_hard");
        }
        return null;
    }

    public void setComplexity(String complexity) {
        if(complexity.equals(Main.locale.getString("easy_short"))) this.complexity = 0;
        else if(complexity.equals(Main.locale.getString("medium_short"))) this.complexity = 1;
        else if(complexity.equals(Main.locale.getString("hard_short"))) this.complexity = 2;
        else if(complexity.equals(Main.locale.getString("very_hard_short"))) this.complexity = 3;
    }

    public void setAddAndColorRow(Boolean sAdd){
        this.add = sAdd;

        if (sAdd) {
            addButton.setVisible(false);
            removeButton.setVisible(true);
            row.getStyleClass().add("isAdd");
        } else {
            addButton.setVisible(true);
            removeButton.setVisible(false);
            row.getStyleClass().remove("isAdd");
        }
    }
}

File: src/main/kotlin/ru/gurps/generator/desktop/models/rules/Spell.kt
package ru.gurps.generator.desktop.models.rules

import ru.gurps.generator.desktop.config.Model
import tornadofx.*
import tornadofx.FX.Companion.messages
import java.sql.ResultSet

class Spell : Model {
    override var id: Int = -1
    var schoolId: Int = -1
    var name: String = ""
    var nameEn: String = ""
    private var _spellType: String = ""
    var spellType: String = ""
        get() = _spellType
                .replace(",", "; ")
                .replace(":", "${messages["or"]}  ")
                .replace("0", "${messages["usual"]}  ")
                .replace("1", "${messages["area"]}  ")
                .replace("2", "${messages["contact"]}  ")
                .replace("3", "${messages["throw"]}  ")
                .replace("4", "${messages["block_spell"]}  ")
                .replace("5", "${messages["resistance"]}  ")
                .replace("6", "${messages["information"]}  ")
                .replace("7", "${messages["charm"]}  ")
                .replace("8", "${messages["special"]}  ")
    var description: String = ""
    var _complexity: Int = -1
    val complexity: String by lazy {
        when (_complexity) {
            0 -> messages["easy"]
            1 -> messages["medium"]
            2 -> messages["hard"]
            3 -> messages["very_hard"]
            else -> ""
        }
    }
    var cost: String = ""
    var needTime: String = ""
    var duration: String = ""
    var maintainingCost: String = ""
    var thing: String = ""
    var createCost: String = ""
    var demands: String = ""
    var resistance: String = ""
    var modifiers: String = ""
    var level: Int = 1
    var finalCost: Int = 0
    var add: Boolean = false
    val school: School by lazy { School().find(schoolId) as School }
    val schoolName: String by lazy { school.name }
    val schoolSingle: String by lazy { "${messages["school"]}: ${school.name}" }
    val resistanceSingle: String by lazy {
        val res = if (resistance.isBlank()) messages["no"] else resistance
        "${messages["resistance"]}: $res"
    }
    val typeSingle: String by lazy { "${messages["type"]}: $spellType" }
    val complexitySingle: String by lazy { "${messages["complexity"]}: $complexity" }
    val demandsSingle: String by lazy {
        val dem = if (demands.isBlank()) messages["no"] else demands
        "${messages["demands"]}: $dem"
    }
    val modifiersSingle: String by lazy { if (modifiers.isBlank()) "" else "${messages["modifiers"]}: $modifiers" }
    val needTimeSingle: String by lazy { "${messages["need_time"]}: $needTime" }
    val costSingle: String by lazy { "${messages["cost"]}: $cost" }
    val maintainingCostSingle: String by lazy { "${messages["maintaining_cost"]}: $maintainingCost" }
    val durationSingle: String by lazy { "${messages["duration"]}: $duration" }
    val thingSingle: String by lazy {
        val thingS = if (thing.isBlank()) messages["no"] else thing
        "${messages["duration"]}: $thingS"
    }
    val createCostSingle: String by lazy {
        val create = if (createCost.isBlank()) messages["no"] else createCost
        "${messages["duration"]}: $create"
    }

    constructor()
    constructor(result: ResultSet) {
        id = result.getInt("id")
        schoolId = result.getInt("schoolId")
        val nameRes = result.getString("name")
        if (nameRes != null) name = nameRes
        val nameEnRes = result.getString("nameEn")
        if (nameEnRes != null) nameEn = nameEnRes
        val spellTypeRes = result.getString("spellType")
        if (spellTypeRes != null) _spellType = spellTypeRes
        val descriptionRes = result.getString("description")
        if (descriptionRes != null) description = descriptionRes
        _complexity = result.getInt("complexity")
        val costRes = result.getString("cost")
        if (costRes != null) cost = costRes
        val needTimeRes = result.getString("needTime")
        if (needTimeRes != null) needTime = needTimeRes
        val durationRes = result.getString("duration")
        if (durationRes != null) duration = durationRes
        val maintainingCostRes = result.getString("maintainingCost")
        if (maintainingCostRes != null) maintainingCost = maintainingCostRes
        val thingRes = result.getString("thing")
        if (thingRes != null) thing = thingRes
        val createCostRes = result.getString("createCost")
        if (createCostRes != null) createCost = createCostRes
        val demandsRes = result.getString("demands")
        if (demandsRes != null) demands = demandsRes
        val resistanceRes = result.getString("resistance")
        if (resistanceRes != null) resistance = resistanceRes
        val modifiersRes = result.getString("modifiers")
        if (modifiersRes != null) modifiers = modifiersRes
    }

    fun setComplexity(complexity: String) {
        when (complexity) {
            messages["easy"] -> this._complexity = 0
            messages["medium"] -> this._complexity = 1
            messages["hard"] -> this._complexity = 2
            messages["very_hard"] -> this._complexity = 3
        }
    }
}
